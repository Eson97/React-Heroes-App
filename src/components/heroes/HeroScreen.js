import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router'
import { heroesImages } from '../../helpers/heroesImages';
import { getHeroesByid } from '../../selectors/getHeroById';

export const HeroScreen = ({ history }) => {

    const { heroeId } = useParams(); //obtiene los argumentos del url

    const hero = useMemo(() => getHeroesByid(heroeId), [heroeId]);

    if (!hero) return <Redirect to="/" />

    const { superhero, publisher, alter_ego, first_appearance, characters } = hero

    const handleReturn = () => {
        (hero.publisher === 'DC Comics')
            ? history.push('/dc')
            : history.push('/marvel')
    }

    return (
        <div className="row mt-5 justify-content-center animate__animated animate__fadeInRight">
            <div className="col-4 ">
                <img
                    className="img-thumbnail"
                    // src={`../assets/heroes/${heroeId}.jpg`} //Obtener imagenes desde la carpeta 'Public'
                    src={heroesImages(`./${heroeId}.jpg`).default} //Obtiene imagenes desde src (.default es necesario para mostrar imagen)
                    alt={superhero}
                />
            </div>

            <div className="w-100 d-sm-none"></div>

            <div className="col-8 mt-3">
                <h3>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"><b>Alter ego: </b>{alter_ego}</li>
                    <li className="list-group-item"><b>Publisher: </b>{publisher}</li>
                    <li className="list-group-item"><b>First Appearance: </b>{first_appearance}</li>
                </ul>
                <h5 className="mt-3">Characters</h5>
                <p>{characters}</p>

                <div className="d-grid gap-2 btn-lg d-md-block">
                    <button className="btn btn-outline-info" onClick={handleReturn}>
                        {
                            (hero.publisher === 'DC Comics')
                                ? "Go to DC Comics"
                                : "Go to Marvel Comics"
                        }
                    </button>
                </div>
            </div>


        </div>
    )
}
