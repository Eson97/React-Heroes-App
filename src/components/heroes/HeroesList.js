import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard';

export const HeroesList = ({ publisher }) => {

    const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher]); //solo se ejecuta si el publisher cambia

    return (
        <div className="row">
            {
                heroes.map(hero => (
                    <div className="col-6 col-md-4 animate__animated animate__fadeIn" key={hero.id}>
                        <HeroCard {...hero} />
                    </div>
                ))
            }
        </div>
    )
}
