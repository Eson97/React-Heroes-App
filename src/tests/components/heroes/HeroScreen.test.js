import { mount, shallow } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { HeroScreen } from "../../../components/heroes/HeroScreen"

describe('Pruebas en HeroScreen', () => {

    const historyMock = {
        lenght: 10,
        push: jest.fn(),
        goBack: jest.fn()
    }


    test('should show comp redirect if there si no args in URL', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={historyMock} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true);
    })

    test('should display a hero if param exist', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroeId" component={HeroScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('.row').exists()).toBeTruthy();
    })

    test('should return to previous screen using PUSH', () => {
        const historyMock = {
            lenght: 10,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroeId" component={() => <HeroScreen history={historyMock} />} />
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')();

        expect(historyMock.push).toHaveBeenCalled();
        expect(historyMock.goBack).not.toHaveBeenCalled();
    })

    test('should call REDIRECT if hero doesnÂ´t exist', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider654654564654']}>
                <Route path="/hero/:heroeId" component={() => <HeroScreen history={historyMock} />} />
            </MemoryRouter>
        )

        expect(wrapper.text()).toBe('');
    })




})
