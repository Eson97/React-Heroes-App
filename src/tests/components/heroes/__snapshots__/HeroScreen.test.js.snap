// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en HeroScreen should show comp redirect if there si no args in URL 1`] = `""`;
